src/main/java/frc/robot/subsystems/drive/states/LinearDriveState.java

```diff
@@ -0,0 +1,266 @@
+                JsonConstants.blueFieldLocations.blueCoralStationLeftRotation);
+      default:
+        return null;
+    }
+  }
+
+  @Override
+  public void periodic() {
+    Pose2d currentPose = drive.getPose();
+
+    double currentDistance = currentPose.getTranslation().getDistance(goalPose.getTranslation());
+
+    if (withinRange(currentPose, goalPose, 0.05)) {
+      drive.setGoalSpeeds(new ChassisSpeeds(), true);
+
+      if (drive.isDesiredLocationReef()) {
+        drive.fireTrigger(DriveTrigger.BeginLineup);
+      } else {
+        drive.fireTrigger(DriveTrigger.CancelLinear);
+      }
+      return;
+    }
+
+    // DRIVE
+    double ffScaler =
+        MathUtil.clamp(
+            (currentDistance - m_ffMinRadius) / (m_ffMaxRadius - m_ffMinRadius), 0.0, 1.0);
+    double driveVelocityScalar =
+        driveController.getSetpoint().velocity * ffScaler
+            + driveController.calculate(currentDistance, 0.0);
+    if (currentDistance < driveController.getPositionTolerance()) {
+      driveVelocityScalar = 0.0;
+    }
+
+    // HEADING
+    double headingError = currentPose.getRotation().minus(goalPose.getRotation()).getRadians();
+    double headingVelocity =
+        headingController.calculate(
+            currentPose.getRotation().getRadians(), goalPose.getRotation().getRadians());
+    if (Math.abs(headingError) < headingController.getPositionTolerance()) {
+      headingVelocity = 0.0;
+    }
+
+    Translation2d driveVelocity =
+        new Pose2d(
+                0.0, 0.0, currentPose.getTranslation().minus(goalPose.getTranslation()).getAngle())
+            .transformBy(new Transform2d(driveVelocityScalar, 0.0, new Rotation2d()))
+            .getTranslation();
+    ChassisSpeeds speeds =
+        ChassisSpeeds.fromFieldRelativeSpeeds(
+            driveVelocity.getX(), driveVelocity.getY(), headingVelocity, currentPose.getRotation());
+    drive.setGoalSpeeds(speeds, true);
+
+    Logger.recordOutput("DriveToPoint/TargetPose", goalPose);
+    Logger.recordOutput("DriveToPoint/DriveDistance", currentDistance);
+    Logger.recordOutput("DriveToPoint/HeadingError", headingError);
+
+    Logger.recordOutput("DriveToPoint/FFScaler", ffScaler);
+    Logger.recordOutput("DriveToPoint/DriveVelocityScalar", driveVelocityScalar);
+    Logger.recordOutput("DriveToPoint/HeadingVelocity", headingVelocity);
+    Logger.recordOutput("DriveToPoint/DriveVelocityX", driveVelocity.getX());
+    Logger.recordOutput("DriveToPoint/DriveVelocityY", driveVelocity.getY());
+
+    Logger.recordOutput("DriveToPoint/DriveSpeeds", speeds);
+  }
+
+  public boolean withinRange(Pose2d a, Pose2d b, double error) {
+    System.out.println(Math.abs(a.getTranslation().getX() - b.getTranslation().getX()));
+    return Math.abs(a.getTranslation().getX() - b.getTranslation().getX()) <= error
+        && Math.abs(a.getTranslation().getY() - b.getTranslation().getY()) <= error
```

Comment:
Rather than doing x and y in range separately, we should get the distance between A and B and see if it's less than error